### PR TITLE
437 delete stale demo account

### LINKED ISSUE #437: H6: Stale demo accounts are deleted.
### Description
Demo accounts will build up over time with cognito and mongo getting filled with unused data. We need a way to clear these out manually or ideally automatically.

---

### Proposed Solution(s)
 A script might be needed here that runs based on a timestamp or heuristic value.

---

### Deliverable(s)
- [ ] Stop stale user accounts from building up

### PR DESCRIPTION
**Pull request by O-Susanoo**
---
daniel:

ready to go
closes #437 

The serverless file stuff I am unsure about - and didn't change anything there from the original PR

### CODE REVIEW

File: lib/models/DemoUser/DemoUser.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+const uuidv4 = require('uuid').v4
+const mixin = require('functional-helpers/mixin')
+const Connectable = require('models/Connectable')
+const deleteAllUserRelatedData = require('tasks/users/deleteAllUserRelatedData')
+const CognitoAPI = require('cognito/CognitoAPI')
+
+class DemoUser {
```
Comment (DanielSeehausen) [inline lib/models/DemoUser/DemoUser.js:7]: this file looks like a great start

I believe this should extend the `User` class and make use of `super` in at least `create`

File: lib/models/DemoUser/DemoUser.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    user.createdAt = timestamp
+    const { insertedCount } = await DemoUser.usersDB.insertOne(user)
+    const success = insertedCount === 1
+    if (success) delete user._id
+    return user
+  }
+
+  static async findUnActiveDemoUsers(timestamp) {
+    return DemoUser.usersDB.find({ createdAt: { $lt: timestamp } }).toArray()
+  }
```
Comment (DanielSeehausen) [inline lib/models/DemoUser/DemoUser.js:30]: once extending the `User` class, you could use `findBy` to reduce direct database access

File: lib/models/DemoUser/DemoUser.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+  static async removeUnActiveUsers(timestamp) {
+    const unActiveDemoUsers = await DemoUser.findUnActiveDemoUsers(timestamp)
+    return Promise.all(unActiveDemoUsers.map(async demoUser => {
+      await deleteAllUserRelatedData(demoUser.id)
+      return demoUser.id
+    }))
+  }
```
Comment (DanielSeehausen) [inline lib/models/DemoUser/DemoUser.js:38]: this is extending a little beyond the scope of the demo user interface (base `User` already has `delete` on it and that should probably be enough once this is extending it). Whatever the service/task is that is doing the clean up, that is what should probably be responsible for doing the `deleteAllUserRelatedData`. The tasks can maybe be best thought of as higher level actions that delegate responsibilities down to the interfaces, instead of the other way around

File: services/user/src/user/createDemoUser.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+const connectAllModels = require('models/connectAllModels')
+const DemoUser = require('models/DemoUser/DemoUser')
+const HTTPDefaultMiddleware = require('middleware/HTTPDefaultMiddleware')
+const { success } = require('http-responses')
+
+async function createDemoUser(event, context) {
+  await connectAllModels()
+  const timestamp = new Date().getTime()
+  const demoUser = DemoUser.create(timestamp)
```
Comment (DanielSeehausen) [inline services/user/src/user/createDemoUser.js:9]: extending the `User` class should free you up from needing to do this iirc
I belie `DemoUser` needs to be awaited

File: test/lib/models/DemoUser/DemoUser.test.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      await DemoUser.create(timestamp)
+      expect(CognitoAPI.createUser).toHaveBeenCalledTimes(1)
+    })
+
+    it('should assign the correct username to the demo account', async () => {
+      const timestamp = 12345
+      // const demoUser = await DemoUser.create(timestamp)
+
+      // expect(demoUser.username).toBe('guest-12345')
+    })
```
Comment (DanielSeehausen) [inline test/lib/models/DemoUser/DemoUser.test.js:36]: this test may not be finished

File: lib/models/DemoUser/DemoUser.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    const userName = `guest-${timestamp}`
+    const user = {
+      id: uuidv4(),
+      role: 'demo',
+      userName,
+      email: `${userName}@ospin.de`,
+      processes: [],
+      devices: [],
+    }
+    await CognitoAPI.createUser(user.userName, password)
```
Comment (DanielSeehausen) [inline lib/models/DemoUser/DemoUser.js:20]: I believe this might be better suited in the lambda function instead of the model interface

File: lib/models/DemoUser/DemoUser.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
+const uuidv4 = require('uuid').v4
+const mixin = require('functional-helpers/mixin')
+const Connectable = require('models/Connectable')
+const deleteAllUserRelatedData = require('tasks/users/deleteAllUserRelatedData')
+const CognitoAPI = require('cognito/CognitoAPI')
+
+class DemoUser {
+
+  static async create(timestamp) {
```
Comment (DanielSeehausen) [inline lib/models/DemoUser/DemoUser.js:9]: timestamp should likely be auto provided by the create method (see `User.create`).

File: lib/models/DemoUser/DemoUser.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    const success = insertedCount === 1
+    if (success) delete user._id
+    return user
+  }
+
+  static async findUnActiveDemoUsers(timestamp) {
+    return DemoUser.usersDB.find({ createdAt: { $lt: timestamp } }).toArray()
+  }
+
+  static async removeUnActiveUsers(timestamp) {
```
Comment (DanielSeehausen) [inline lib/models/DemoUser/DemoUser.js:32]: to make this a little more clear with an argument, there are two ways we could potentially refine this:
a. make a static getter to define what duration in ms is considered a 'stale/inactive' user and reference it in this method
b. edit the method name to communicate the use case, potentially something like `removeUsersInactiveForMs`

File: test/lib/models/DemoUser/DemoUser.test.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+  })
+
+  describe('when demo user in database for more than one month', () => {
```
Comment (DanielSeehausen) [inline test/lib/models/DemoUser/DemoUser.test.js:40]: this test might be a little misleading with the description and what it tests for

File: test/lib/models/DemoUser/DemoUser.test.js [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    it('should remove demo user from cognito pool', async () => {
+      await DemoUser.create(1000)
```
Comment (DanielSeehausen) [inline test/lib/models/DemoUser/DemoUser.test.js:42]: requested above that `create` extend and delegate to `User.create`, which means the create time argument will unfortunately be lost. The `User.update` method could be used instead to force a creation date in the past.

File: lib/models/DemoUser/DemoUser.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+      id: uuidv4(),
+      role: 'demo',
+      userName,
+      email: `${userName}@ospin.de`,
+      processes: [],
+      devices: [],
+    }
+    await CognitoAPI.createUser(user.userName, password)
+    user.createdAt = timestamp
+    const { insertedCount } = await DemoUser.usersDB.insertOne(user)
```
Comment (DanielSeehausen) [inline lib/models/DemoUser/DemoUser.js:22]: one benefit of extending the `User` class is to ensure data gets pushed through the user schema before it is persisted

File: services/user/src/user/create.js [hunk 12]
```diff
@@ -19,9 +19,10 @@ async function create(event, context) {
     email: event.request.userAttributes.email,
     phoneNumber: event.request.userAttributes.phoneNumber,
     userName: event.userName,
-    role: 'user', // currently restricted to just 'user'
+    role: event.role || 'user', // currently restricted to just 'user'
```
Comment (DanielSeehausen) [inline services/user/src/user/create.js:22]: seeing as the new endpoint `createDemoUser` is here now, is this change still required?

File: lib/models/User/User.js [hunk 13]
```diff
@@ -53,9 +53,7 @@ class User {
   }
 
   static async find(userId) {
-    const user = await User.usersDB.findOne({ id: userId })
-    if (user) delete user._id
-
+    const user = await User.findBy({ id: userId })
```
Comment (vdiezel) [inline lib/models/User/User.js:56]: `findById` ?

File: services/user/src/user/jobs/removeUnActiveDemoUsers.js [hunk 14]
```diff
@@ -0,0 +1,11 @@
+const connectAllModels = require('models/connectAllModels')
+const DemoUser = require('models/DemoUser/DemoUser')
+
+async function removeUnActiveUsers() {
+  await connectAllModels()
+  const now = new Date()
+  now.setMonth(now.getMonth() - 1)
```
Comment (vdiezel) [inline services/user/src/user/jobs/removeUnActiveDemoUsers.js:7]: very cool syntax

**Comment by DanielSeehausen** [comment]
closes #437 

big update - probably easiest to review anew!

File: services/user/test/user/jobs/removeInactiveDemoUsers.test.js [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+      staleUpdatedAndCreatedDemoUserA,
+      staleUpdatedAndCreatedDemoUserB,
+    } = await getVarietyOfUsers()
+
+    const { deletedIds } = await removeInactiveDemoUsers()
+
+    expect(deleteAllUserRelatedData).toHaveBeenCalledTimes(2)
+    expect(deleteAllUserRelatedData).toHaveBeenNthCalledWith(1, staleUpdatedAndCreatedDemoUserA.id)
+    expect(deleteAllUserRelatedData).toHaveBeenNthCalledWith(2, staleUpdatedAndCreatedDemoUserB.id)
+
```
Comment (vdiezel) [inline services/user/test/user/jobs/removeInactiveDemoUsers.test.js:92]: an assertion for the call to the CognitoAPI would  be nice

File: services/user/src/user/jobs/removeInactiveDemoUsers.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
+const connectAllModels = require('models/connectAllModels')
+const deleteAllUserRelatedData = require('tasks/users/deleteAllUserRelatedData')
+const DemoUser = require('models/DemoUser/DemoUser')
```
Comment (vdiezel) [inline services/user/src/user/jobs/removeInactiveDemoUsers.js:3]: there is no DemoUser directory in models - its in User in my case

File: services/user/src/user/jobs/removeInactiveDemoUsers.js [hunk 17]
```diff
@@ -1,6 +1,6 @@
 const connectAllModels = require('models/connectAllModels')
 const deleteAllUserRelatedData = require('tasks/users/deleteAllUserRelatedData')
-const DemoUser = require('models/DemoUser/DemoUser')
+const DemoUser = require('models/User/DemoUser')
```
Comment (ConorTighe1995) [inline services/user/src/user/jobs/removeInactiveDemoUsers.js:3]: Nice one viktor was about to push this
